source/mainnet/smart-contracts/best-practices/development.rst

```diff
@@ -0,0 +1,203 @@
+A procedure can be interrupted in the middle of its execution, run again in *another* execution context, and then continue execution from the interruption point.
+In case of smart contracts, each external call interrupts the execution and hands over control to unknown code.
+Do not treat external contract invocations as regular method calls.
+Instead, think of them as sending a message and temporarily pausing execution of your contract.
+The receiving side has full control of what to do next and can choose to call your contract again while it is still in the "paused" state waiting for the external call to be completed.
+Once the external call is completed, you might find your contract in a completely different state.
+See an :ref:`example <reentracny-unit-testing>`, based on the DAO contract vulnerability of how reentrancy can be discovered using unit testing.
+
+- Avoid changing the state after an external call: use the *Checks-Effects-Interactions* pattern: validate data, update the contract state, make external calls.
+- If you need to perform some state changes after an external call use `invoke_contract_read_only <https://docs.rs/concordium-std/latest/concordium_std/trait.HasHost.html#method.invoke_contract_read_only>`_.
+  If the read-only invocation succeeds, it ensures that the state has not been changed after returning from the external call.
+  Using ``invoke_contract`` covers most of the cases where protection from reentrancy is required.
+- Alternatively, consider using a *mutex*: a boolean flag that is set before making an external call, preventing all entrypoints from reentrancy. Reset after the call is complete.
+
+.. _best-practices-code-documentation:
+
+Code documentation
+------------------
+
+- Write an outline of the smart contract functionality in the beginning of the file; if the contract implements some standards, mention it.
+- Document decisions/choices in the code.
+- Document entrypoints:
+
+  - What functionality the entrypoint implements?
+  - Who has access rights to call the entrypoint?
+  - When the call is rejected?
+  - What events are logged?
+
+- Document tests:
+
+  - What scenario/property is being tested?
+  - What are the assumptions: input data is assumed to be valid, users have enough rights, etc.
+
+.. _best-practices-code-automated-testing:
+
+Automated testing
+-----------------
+
+The Concordium standard library `concordium-std`_ offers several possibilities for testing the smart contract code.
+
+- :ref:`Unit testing <unit-test-contract>` used for testing particular cases where you define what is the valid output.
+- :ref:`Property-based testing <writing_property_based_tests>` is a variant of randomized testing that repeatedly checks a *property* with randomly generated input.
+
+Use the :ref:`smart contract specification <best-practices-specification>` to come up with cases and properties to test.
```

Comment:
It wasn't clear to me that this referred to a section in this page.

Reply:
I don't fully understand what you mean here. Do you mean there is another way of referencing the content on the same page? Could you please clarify?

Reply:
When I read it, I thought it referred to some other page or guideline and then I was briefly confused about it leading to a section on the same page.

I would just add a few more words, e.g.
```suggestion
Use the :ref:`smart contract specification <best-practices-specification>` guidelines from the section above to come up with cases and properties to test.
```